File: codes/python/chap08/answer_A58.py
# 세그멘트 트리(여기에서는 책과 달리, pos가 0-indexed가 되도록 구현했습니다)
class segtree:
	# 요소 dat의 초기화를 수행한다(최초에는 전부 0)
	def __init__(self, n):
		self.size = 1
		while self.size < n:
			self.size *= 2
		self.dat = [ 0 ] * (self.size * 2)

  # 쿼리 1에 대한 처리	
	def update(self, pos, x):
		pos += self.size # pos는 0-indexed이므로, A[pos]에만 대응하는 셀의 번호는 pos + size
		self.dat[pos] = x
		while pos >= 2:
			pos //= 2
			self.dat[pos] = max(self.dat[pos * 2], self.dat[pos * 2 + 1])
	
  # 쿼리 2에 대한 처리
  # u는 현재의 셀 번호, [a, b)는 셀에 대응하는 반개구간, [l, r)은 구하는 반개구간
	def query(self, l, r, a, b, u):
		if r <= a or b <= l:
			return -1000000000 # 일절 포함되지 않는 경우
		if l <= a and b <= r:
			return self.dat[u] # 완전히 포함되는 경우
		m = (a + b) // 2
		answerl = self.query(l, r, a, m, u * 2)
		answerr = self.query(l, r, m, b, u * 2 + 1)
		return max(answerl, answerr)

# 입력
N, Q = map(int, input().split())
queries = [ list(map(int, input().split())) for i in range(Q) ]

# 쿼리 처리
Z = segtree(N)
for q in queries:
	tp, *cont = q
	if tp == 1:
		pos, x = cont
		Z.update(pos - 1, x) # pos는 1-indexed로 입력되어 있으므로, update 함수의 인수는 pos - 1로 합니다
	if tp == 2:
		l, r = cont
		answer = Z.query(l - 1, r - 1, 0, Z.size, 1) # 0-indexed의 구현에서는, 최초의 셀에 대응하는 반개구간은 [0, size)입니다
		print(answer)

File: codes/cpp/chap08/answer_A58.cpp
#include <iostream>
#include <algorithm>
using namespace std;

class SegmentTree {
public:
	int dat[300000], siz = 1;

	// 요소 dat의 초기화를 수행한다(최초에는 전부 0)
	void init(int N) {
		siz = 1;
		while (siz < N) siz *= 2;
		for (int i = 1; i < siz * 2; i++) dat[i] = 0;
	}

	// 쿼리 1에 대한 처리
	void update(int pos, int x) {
		pos = pos + siz - 1;
		dat[pos] = x;
		while (pos >= 2) {
			pos /= 2;
			dat[pos] = max(dat[pos * 2], dat[pos * 2 + 1]);
		}
	}

	// 쿼리 2에 대한 처리
	// u는 현재의 셀 번호, [a, b)는 셀에 대응하는 반개구간, [l, r)은 구할 반개구간
	int query(int l, int r, int a, int b, int u) {
		if (r <= a || b <= l) return -1000000000; // 전혀 포함되지 않는 경우
		if (l <= a && b <= r) return dat[u]; // 온전히 포함된 경우
		int m = (a + b) / 2;
		int AnswerL = query(l, r, a, m, u * 2);
		int AnswerR = query(l, r, m, b, u * 2 + 1);
		return max(AnswerL, AnswerR);
	}
};

int N, Q;
int Query[100009], pos[100009], x[100009], l[100009], r[100009];
SegmentTree Z;

int main() {
	// 입력
	cin >> N >> Q;
	for (int i = 1; i <= Q; i++) {
		cin >> Query[i];
		if (Query[i] == 1) cin >> pos[i] >> x[i];
		if (Query[i] == 2) cin >> l[i] >> r[i];
	}

	// クエリ処理
	Z.init(N);
	for (int i = 1; i <= Q; i++) {
		if (Query[i] == 1) {
			Z.update(pos[i], x[i]);
		}
		if (Query[i] == 2) {
			// 最初のセルに対応する半開区間は [1, siz + 1)
			int Answer = Z.query(l[i], r[i], 1, Z.siz + 1, 1);
			cout << Answer << endl;
		}
	}
	return 0;
}
